src/care/LoopFuser.cpp

```diff
@@ -151,194 +135,94 @@ void LoopFuser::warnIfNotFlushed() {
    }
 }
 
-void LoopFuser::flush_parallel_actions(bool async) {
+void LoopFuser::flush_parallel_actions(bool async, const char * fileName, int lineNumber) {
    // Do the thing
-#ifdef FUSER_VERBOSE
-   if (m_verbose) {
-      printf("in flush_parallel_actions with %i,%i\n", m_action_count, m_max_action_length);
-   }
-#endif
-   SerializableDeviceLambda<int> *actions = m_actions;
-   int * offsets = m_action_offsets;
-
-   int end = m_action_offsets[m_action_count-1];
-   int action_count = m_action_count;
-#ifdef FUSER_VERBOSE
-   if (m_verbose) {
-      printf("launching with end %i and action_count %i\n", end, action_count);
+   if (verbose) {
+      printf("in flush_parallel_actions at %s:%i with %zu, %i\n", fileName, lineNumber, m_actions.num_loops(), m_max_action_length);
    }
-#endif
-   bool reverse_indices = m_reverse_indices;
-   CARE_STREAM_LOOP(i, 0, end) {
-      int index = i;
-      if (reverse_indices) {
-         // do indices in reverse order to discover any order dependencies between loops
-         // and possibly therefore debug GPU race conditions in debug CPU builds of the code.
-         index = end-1-i;
-      }
-      int actionIndex = care::binarySearch<int>(offsets, 0, action_count, index, true);
-#ifdef FUSER_VERBOSE
-      if (m_verbose) {
-         printf("launching action %i with index %i\n", actionIndex, index);
-      }
-#endif
-      actions[actionIndex](index, true, actionIndex, -1, -1);
-   } CARE_STREAM_LOOP_END
-   if (!async) {
-#ifdef FUSER_VERBOSE
-      if (m_verbose) {
-         printf("syncing \n");
-      }
-#endif
-      care::syncIfNeeded();
+   action_workgroup aw = m_actions.instantiate();
+   action_worksite aws = aw.run(nullptr, fusible_registers{});
+   // this resets m_conditionals, which we will never need to run
+   m_conditionals.clear();
+   if (verbose) {
+      printf("done with flush_parallel_actions at %s:%i with %zu, %i, async %i\n", fileName, lineNumber, m_actions.num_loops(), m_max_action_length, (int) async);
    }
+   reset(async, fileName, lineNumber);
```

Comment:
If this is asynchronous then it looks like the workgroup and worksite which store per instantiation and per run data will be destroyed and release said data before the run is complete.